### PR TITLE
GTM: Add AMP container select

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #413 

## Relevant technical choices

* Updated `GET:tag-permission` to be usage context agnostic
* Updated `GET:container` to accept an array of `usageContext`s
* Updated `Module::exception_to_error` from private to protected
    - Allows the method to be used from Module datapoint callbacks
    - Allowed for some module-specific logic (AdSense) to be moved to that module rather than in the base class
    - Used by `POST:settings` now as part of larger simplification around container creation logic
* Updated `GET:containers` to not automatically create a container if the account has none
    - Updated setup interface to automatically select `create_container` if no containers exist
* Updated E2E tests for setup to ensure AMP containers are not listed in web container select
* Updated labelling of containers in settings and in setup UI
    - Container select is labeled "Container" unless both selects are shown
* Updated settings to show "Not set" instead of `false` when setting is not set yet
* Tweaked language around snippet insertion to "on your site" rather than "into" or "to your site"
* Update container name generated for AMP container to prevent naming conflict

**No AMP**
image

**AMP Secondary**
image

image

**AMP Primary**
image

**Settings**
image

Always displays items for both container IDs

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/tagmanager/setup.js [hunk 1]
```diff
@@ -40,38 +40,46 @@ import { addFilter, removeFilter } from '@wordpress/hooks';
...
-		const containerKey = usageContext === 'amp' ? 'ampContainerID' : 'containerID';
+		const ampUsageContext = ampMode === 'primary' ? USAGE_CONTEXT_AMP : [ USAGE_CONTEXT_WEB, USAGE_CONTEXT_AMP ];
 
 		this.state = {
+			ampEnabled,
+			ampMode,
 			isLoading: true,
 			accounts: [],
-			containers: [],
+			containersWeb: [],
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: I don't think we need to do this renaming. "Web" is the standard here. "AMP" needs to be flagged, but web is overly verbose. I get it for the distinction, but it's different from how we do it on the server. And in a way it clutters this PR quite a bit too :)

Let's leave this as `containers`.

File: assets/js/modules/tagmanager/setup.js [hunk 2]
```diff
@@ -40,38 +40,46 @@ import { addFilter, removeFilter } from '@wordpress/hooks';
...
+			containersAMP: [],
 			errorCode: false,
 			errorMsg: '',
 			selectedAccount: settings.accountID,
-			selectedContainer: settings[ containerKey ],
+			selectedContainerWeb: settings.containerID,
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: Same here, let's leave this `selectedContainer`.

File: assets/js/modules/tagmanager/setup.js [hunk 3]
```diff
@@ -151,15 +159,17 @@ class TagmanagerSetup extends Component {
 			// Verify the user has access to existing tag if found.
 			try {
 				const { account, container } = await data.get( TYPE_MODULES, 'tagmanager', 'tag-permission', { tag: existingContainerID } );
+				const containersWeb = getContainers( [ container ] ).byContext( USAGE_CONTEXT_WEB );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: This is just an internal variable, but the same thought would apply here.

File: assets/js/modules/tagmanager/setup.js [hunk 4]
```diff
@@ -436,15 +443,25 @@ class TagmanagerSetup extends Component {
...
 						</p>
 						<h5 className="googlesitekit-settings-module__meta-item-data">
-							{ accountID || false }
+							{ accountID || __( 'Not set', 'google-site-kit' ) }
+						</h5>
+					</div>
+
+					<div className="googlesitekit-settings-module__meta-item">
+						<p className="googlesitekit-settings-module__meta-item-type">
+							{ __( 'Web Container ID', 'google-site-kit' ) }
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: Good point specifying that only if both container dropdowns are there.

File: assets/js/modules/tagmanager/setup.js [hunk 5]
```diff
@@ -436,15 +443,25 @@ class TagmanagerSetup extends Component {
...
+						</p>
+						<h5 className="googlesitekit-settings-module__meta-item-data">
+							{ settings.containerID || __( 'Not set', 'google-site-kit' ) }
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: Not sure about the "Not set" here, we don't do this anywhere else. Maybe just `false` for now like in most other cases?

File: assets/js/modules/tagmanager/setup.js [hunk 6]
```diff
@@ -436,15 +443,25 @@ class TagmanagerSetup extends Component {
...
 					</div>
+
 					<div className="googlesitekit-settings-module__meta-item">
 						<p className="googlesitekit-settings-module__meta-item-type">
-							{ __( 'Container ID', 'google-site-kit' ) }
+							{ __( 'AMP Container ID', 'google-site-kit' ) }
 						</p>
 						<h5 className="googlesitekit-settings-module__meta-item-data">
-							{ settings[ containerKey ] || false }
+							{ settings.ampContainerID || __( 'Not set', 'google-site-kit' ) }
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: This should only display if `ampMode` is `secondary`.

File: assets/js/modules/tagmanager/setup.js (hunk 5 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: Isn't `false` just blank though (i.e. an empty string)? My thought was that an "empty" value would be more useful than a blank value. To your point though, if we were to change this it should probably be done for all modules at once rather than in one place. I'll update this for now but let me know if you'd like me to open an issue to discuss this.

File: assets/js/modules/tagmanager/setup.js (hunk 1 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: I think the naming change makes it easier to understand by removing the need to infer whether `containers` refers to web containers or a collection that could potentially contain both. I understand this adds a few lines to the diff, but I think they're worth it.

There are some cases where this naming separation needs to remain to avoid making things more complicated by reassigning variables, etc; (e.g. `requestTagManagerAccountsContainers`) I assume those are okay to keep?

File: assets/js/modules/tagmanager/setup.js [hunk 7]
```diff
@@ -228,17 +242,26 @@ class TagmanagerSetup extends Component {
...
+			// If the selected container is not in the list of containers, clear it.
+			const containerIDs = containers.map( ( { publicId } ) => publicId ); /* Capitalization rule exception: `publicId` is a property of an API returned value. */
+			if ( isValidContainerID( selectedContainer ) && ! containerIDs.includes( selectedContainer ) ) {
+				selectedContainer = '';
 			}
+			if ( isValidContainerID( selectedContainerAMP ) && ! containerIDs.includes( selectedContainerAMP ) ) {
+				selectedContainerAMP = '';
+			}
+
+			const containersWeb = getContainers( containers ).byContext( USAGE_CONTEXT_WEB );
```
Comment (aaemnnosttv) [inline assets/js/modules/tagmanager/setup.js]: This is named differently to not conflict with `containers` which is already defined in the current scope.

File: includes/Modules/Tag_Manager.php [hunk 8]
```diff
@@ -498,35 +498,23 @@ protected function create_data_request( Data_Request $data ) {
...
-				$usage_context = $data['usageContext'] ?: self::USAGE_CONTEXT_WEB;
-
-				if ( self::USAGE_CONTEXT_WEB === $usage_context && ! isset( $data['containerID'] ) ) {
-					/* translators: %s: Missing parameter name */
-					return new WP_Error( 'missing_required_param', sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), 'containerID' ), array( 'status' => 400 ) );
-				}
-				if ( self::USAGE_CONTEXT_AMP === $usage_context && ! isset( $data['ampContainerID'] ) ) {
-					/* translators: %s: Missing parameter name */
-					return new WP_Error( 'missing_required_param', sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), 'ampContainerID' ), array( 'status' => 400 ) );
-				}
```
Comment (felixarntz) [inline includes/Modules/Tag_Manager.php]: Why was this removed? We should still catch these errors I think.
* If secondary AMP, require both.
* If primary AMP, require `ampContainerID`.
* If no AMP, require `containerID`.

This should maybe depend on the passed `usageContext` instead of `Context` - what do you think?

File: includes/Modules/Tag_Manager.php [hunk 9]
```diff
@@ -566,38 +554,31 @@ protected function create_data_request( Data_Request $data ) {
...
-			$restore_defer();
-			$message = $e->getErrors();
-			if ( isset( $message[0]['message'] ) ) {
-				$message = $message[0]['message'];
-			}
-			return new WP_Error( $e->getCode(), $message );
-		} catch ( Exception $e ) {
-			$restore_defer();
-			return new WP_Error( $e->getCode(), $e->getMessage() );
-		}
```
Comment (felixarntz) [inline includes/Modules/Tag_Manager.php]: I'm not sure we should remove this try-catch from here. With the updated code, we don't `$restore_defer()` any more when an exception happens, which could become problematic. We should keep the try-catch to do that, we could just re-throw the exception.

File: includes/Modules/Tag_Manager.php (hunk 8 above)
Comment (aaemnnosttv) [reply includes/Modules/Tag_Manager.php]: I think I removed it when removing the conditional in the `data.set()` call which used to be dynamic based on `containerKey`. Now it passes both in all cases. I don't think there's a need to over-complicate things on the JS side to prevent sending container IDs based on usage context.

I updated the datapoint handler to require the necessary parameters based on `usageContext` which is also now required.

File: includes/Modules/Tag_Manager.php (hunk 9 above)
Comment (aaemnnosttv) [reply includes/Modules/Tag_Manager.php]: Good catch! Updated 👍

**Comment by aaemnnosttv** [comment]
Feedback addressed! Ready for another pass 😄
